File: MathBitManipulation/trailingZeros.py
'''
2x5会使末尾产生 0
质因数分解后而 2 的个数一定大于 5 的个数
所以统计 5 的个数即可
例如 11 要统计 11！中会分解出几个 5，可以用 11//5 
但是如果是 25 则多出一个 0，所以对于 //25 要再加一个 0
以此类推 125 再加一个 0
'''

class Solution:
    """
    @param: n: An integer
    @return: An integer, denote the number of trailing zeros in n!
    """
    def trailingZeros(self, n):
        # write your code here, try to do it without arithmetic operators.
        res = 0
        num = 5
        while n >= num:
            res += n // num
            num *= 5
        return res
            
        
        
if __name__ == '__main__':
    solution = Solution()
    print(solution.trailingZeros(21))

File: Freestyle/trailingZeros.cc
/*
Write an algorithm which computes the number of trailing zeros in n factorial.

Example
11! = 39916800, so the out should be 2

Challenge
O(log N) time
*/
/*
搜索1~n中因数有5的个数（有2的个数一定大于等于有5的个数）
结果是 n//5 + n//25 + n//125 + ...
*/


#include <iostream>

class Solution {
public:
    /*
     * @param n: A long integer
     * @return: An integer, denote the number of trailing zeros in n!
     */
    long long trailingZeros(long long n) {
        // write your code here, try to do it without arithmetic operators.
		long long f = 5;
		long long s = 0;
		while (f <= n) {
			s += (n / f);
			f *= 5;
		}
		return s;
    }
};

int main()
{
	using namespace std;
	Solution s;
	int z = s.trailingZeros(5555550000000);
	cout << z << endl;
	return 0;
}